web/src/js/ducks/ui/flow.js

```diff
@@ -87,10 +98,27 @@ export function updateEdit(update) {
     return { type: UPDATE_EDIT, update }
 }
 
-export function stopEdit(flow) {
+export function uploadContent(flow, content, type){
+    return (dispatch) => {
+        dispatch(flowsActions.updateContent(flow, content, type))
+        dispatch({ type: UPLOAD_CONTENT })
+    }
+}
+
+export function stopEdit(flow, modified_flow) {
+    let diff = getDiff(flow, modified_flow)
     return (dispatch) => {
-        dispatch(flowsActions.update(flow, flow)).then(() => {
-            dispatch(flowsActions.updateFlow(flow))
+        if (diff.response && diff.response.content) {
+            dispatch(flowsActions.updateContent(flow, diff.response.content, "response"))
+            delete diff.response.content
+        }
+        if (diff.request && diff.request.content) {
+            dispatch(flowsActions.updateContent(flow, diff.request.content, "request"))
+            delete diff.request.content
+        }
+
+        dispatch(flowsActions.update(flow, diff)).then(() => {
+            dispatch(flowsActions.updateFlow(modified_flow))
```

Comment:
We have multiple dispatch promises here. We should wait for them all (`Promise.all`) before emitting updates :smiley: